api/src/main/java/org/infinispan/api/async/Cache.java

```diff
@@ -0,0 +1,128 @@
+package org.infinispan.api.async;
+
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+
+import org.infinispan.api.common.AsyncIterator;
+import org.infinispan.api.common.KeyValueEntry;
+import org.infinispan.api.common.events.KeyValueListener;
+
+/**
+ * @author Tristan Tarrant &lt;tristan@infinispan.org&gt;
+ * @since 13.0
+ **/
+public interface Cache<K, V> {
+   /**
+    * Get the value of the Key if such exists
+    *
+    * @param key
+    * @return the value
+    */
+   CompletionStage<V> get(K key);
+
+   /**
+    * Insert the key/value if such key does not exist
+    *
+    * @param key
+    * @param value
+    * @return Void
+    */
+   CompletionStage<Boolean> putIfAbsent(K key, V value);
+
+   /**
+    * Save the key/value. If the key exists will replace the value
+    *
+    * @param key
+    * @param value
+    * @return Void
+    */
+   CompletionStage<Void> put(K key, V value);
```

Comment:
`CompletionStage<V>`?

Reply:
We either provide a `V put(K, V)` paired with a `void set(K, V)` or something like JCache which implies the behaviour in the method: `V getAndPut(K, V)` and `void put(K, V)`

Reply:
> `V put(K, V)` paired with a `void set(K, V)`

That would be my preference.

Reply:
Done